### PR TITLE
IDEMPIERE-4787 : Dashboard / Report : add Refresh button

### PR DESCRIPTION
**Pull request by nmicoud**

### CODE REVIEW

File: org.adempiere.ui.zk/WEB-INF/src/org/adempiere/webui/desktop/DashboardController.java [hunk 1]
```diff
@@ -436,6 +436,15 @@ public  boolean render(Component content, MDashboardContent dc, DashboardRunnabl
 				btn.setLabel(Msg.getMsg(Env.getCtx(), "ViewReportInNewTab"));
 				toolbar.appendChild(new Separator("vertical"));
 				toolbar.appendChild(btn);
+
+				btn = new ToolBarButton();
+				btn.setAttribute("Refresh", true);
+				btn.setAttribute("AD_Process_ID", AD_Process_ID);
+				btn.setAttribute("ProcessParameters", processParameters);
+				btn.setAttribute("Parent", content);
+				btn.setImage(ThemeManager.getThemeResource("images/Refresh16.png"));
```
Comment (hengsin) [inline org.adempiere.ui.zk/WEB-INF/src/org/adempiere/webui/desktop/DashboardController.java:445]: need to cater for css font icon here, i.e if (ThemeManager.isUseFontIconForImage()) btn.setIconSclass(...) else btn.setImage(...). There are plenty of examples in the code.

File: org.adempiere.ui.zk/WEB-INF/src/org/adempiere/webui/desktop/DashboardController.java [hunk 2]
```diff
@@ -436,6 +436,15 @@ public  boolean render(Component content, MDashboardContent dc, DashboardRunnabl
...
+				btn.addEventListener(Events.ON_CLICK, this);
```
Comment (hengsin) [inline org.adempiere.ui.zk/WEB-INF/src/org/adempiere/webui/desktop/DashboardController.java:446]: use btn.addEventListener(Events.ON_CLICK, e -> embedReport(content, AD_Process_ID, processParameters, false)); and you can remove many of your changes :)

File: org.adempiere.ui.zk/WEB-INF/src/org/adempiere/webui/desktop/DashboardController.java [hunk 3]
```diff
@@ -581,8 +590,14 @@ else if (btn.getAttribute("AD_Process_ID") != null)
...
-            		if (processId > 0)
-            			openReportInViewer(processId, parameters);
+            		if (processId > 0) {
+            			if (btn.getAttribute("Refresh") != null) {
+            				Component parent = (Component) btn.getAttribute("Parent");
+            				embedReport(parent, processId, parameters, true);
+            			}
+            			else
+            				openReportInViewer(processId, parameters);
+            		}
```
Comment (hengsin) [inline org.adempiere.ui.zk/WEB-INF/src/org/adempiere/webui/desktop/DashboardController.java:600]: this change is not needed if you use the btn.addEventListener changes suggested above.

File: org.adempiere.ui.zk/WEB-INF/src/org/adempiere/webui/desktop/DashboardController.java [hunk 4]
```diff
@@ -422,7 +422,7 @@ public  boolean render(Component content, MDashboardContent dc, DashboardRunnabl
 			if (dc.isEmbedReportContent()) 
 			{
 				String processParameters = dc.getProcessParameters();
-				embedReport(content, AD_Process_ID, processParameters);
+				embedReport(content, AD_Process_ID, processParameters, false);
```
Comment (hengsin) [inline org.adempiere.ui.zk/WEB-INF/src/org/adempiere/webui/desktop/DashboardController.java:425]: one way to reduce the changes further is to remove the isrefresh parameter from embedreport(..) method and grab the iframe reference here. after the embedreport() call, iframe will be the last children of "content". With the iframe local variable added here, the btn event listener below can be simplify further to btn.addEventListener(Events.ON_Click, e -> iframe.setContent(generateReport(AD_Process_ID, parameters)))

File: org.adempiere.ui.zk/WEB-INF/src/org/adempiere/webui/desktop/DashboardController.java (hunk 1 above)
Comment (nmicoud) [reply org.adempiere.ui.zk/WEB-INF/src/org/adempiere/webui/desktop/DashboardController.java]: done

**Comment by nmicoud** [comment]
hi @hengsin 
new commit following your approach ; I never tried the "new" addEventListener method, thanks !

**Review by hengsin — APPROVED** [review]
looks good

**Pull request merged**